src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -950,20 +968,647 @@ void Chttp2ServerListener::Orphan() {
+      const AcceptorPtr acceptor_;
+      grpc_pollset_set* const interested_parties_;
+      Timestamp const deadline_;
+      OrphanablePtr<grpc_endpoint> endpoint_;
+      // Following fields are protected by WorkSerializer.
+      RefCountedPtr<HandshakeManager> handshake_mgr_;
+      // State for enforcing handshake timeout on receiving HTTP/2 settings.
+      absl::optional<EventEngine::TaskHandle> timer_handle_;
+      grpc_closure on_receive_settings_;
+    };
+
+    NewActiveConnection(RefCountedPtr<NewChttp2ServerListener> listener,
+                        grpc_pollset* accepting_pollset, AcceptorPtr acceptor,
+                        const ChannelArgs& args, MemoryOwner memory_owner,
+                        OrphanablePtr<grpc_endpoint> endpoint);
+    void Start(const ChannelArgs& args);
+
+    void Orphan() override;
+
+    // Needed to be able to grab an external weak ref in
+    // NewChttp2ServerListener::OnAccept()
+    using InternallyRefCounted<LogicalConnection>::RefAsSubclass;
+
+   private:
+    bool SendGoAwayImplLocked() override;
+    void DisconnectImmediatelyImplLocked() override;
+    static void OnClose(void* arg, grpc_error_handle error);
+    void OnDrainGraceTimeExpiry() ABSL_LOCKS_EXCLUDED(&mu_);
+
+    RefCountedPtr<NewChttp2ServerListener> const listener_;
+    // FOllowing fields are protected by WorkSerializer.
+    // Set by HandshakingState before the handshaking begins and reset when
+    // handshaking is done.
+    OrphanablePtr<HandshakingState> handshaking_state_;
+    // Set by HandshakingState when handshaking is done and a valid transport
+    // is created.
+    RefCountedPtr<grpc_chttp2_transport> transport_ = nullptr;
+    grpc_closure on_close_;
+    bool shutdown_ = false;
+  };
+
+  // To allow access to RefCounted<> like interface.
+  friend class RefCountedPtr<NewChttp2ServerListener>;
+
+  // Should only be called once so as to start the TCP server. This should only
+  // be called by the config fetcher.
+  void Start() override;
+
+  static void OnAccept(void* arg, grpc_endpoint* tcp,
+                       grpc_pollset* accepting_pollset,
+                       grpc_tcp_server_acceptor* acceptor);
+
+  static void TcpServerShutdownComplete(void* arg, grpc_error_handle error);
+
+  static void DestroyListener(Server* /*server*/, void* arg,
+                              grpc_closure* destroy_done);
+
+  grpc_tcp_server* tcp_server_ = nullptr;
+  grpc_resolved_address resolved_address_;
+  Server::ListenerWrapper* listener_wrapper_ = nullptr;
+  ChannelArgs args_;
+  Mutex mu_;
+  bool add_port_on_start_ ABSL_GUARDED_BY(mu_) = false;
+  // Signals whether the application has triggered shutdown.
+  bool shutdown_ ABSL_GUARDED_BY(mu_) = false;
+  grpc_closure tcp_server_shutdown_complete_ ABSL_GUARDED_BY(mu_);
+  grpc_closure* on_destroy_done_ ABSL_GUARDED_BY(mu_) = nullptr;
+  RefCountedPtr<channelz::ListenSocketNode> channelz_listen_socket_;
+  MemoryQuotaRefPtr memory_quota_;
```

Comment:
Do these two quota objects belong here, or should they be in the transport-agnostic `ListenerWrapper` code?